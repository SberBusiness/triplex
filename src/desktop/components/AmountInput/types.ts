import {INumberBaseInputProps} from '../protected/NumberBaseInput/types';

/**
 * Свойства компонента AmountInput.
 *
 * @extends INumberBaseInputProps
 */
export interface IAmountInputProps extends Omit<INumberBaseInputProps, 'valueFractionLength' | 'placeholderFractionLength'> {
    /**
     * @prop {string} currency Валюта для инпута (USD, RUB).
     */
    currency: string;
    /**
     * @prop {number} [fractionLength] количество знаков после разделителя.
     */
    fractionLength?: number;
}
