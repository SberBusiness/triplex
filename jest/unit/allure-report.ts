import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
