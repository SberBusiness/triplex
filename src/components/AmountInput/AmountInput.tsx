import React, {useRef} from 'react';
import {AmountBaseInput} from '../protected/AmountBaseInput/AmountBaseInput';
import {IInputProps} from '@sberbusiness/triplex/components/Input/Input';
import {TestProps} from '@sberbusiness/triplex/types/CoreTypes';
import {EInputGroupPosition} from '@sberbusiness/triplex/components/InputGroup/InputGroup';
import {Text} from '@sberbusiness/triplex/components/Typography/Text';
import {EFontType, ETextSize} from '@sberbusiness/triplex/components/Typography/enums';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Свойства компонента AmountInput. */
export interface IAmountInputProps extends Omit<IInputProps, 'type' | 'onChange'>, TestProps {
    value: string;
    onChange: (value: string) => void;
    /** Количество знаков после запятой. */
    fractionLength?: number;
    /** Валюта. */
    currency?: string;
    /** Позиция внутри компонента InputGroup. */
    groupPosition?: EInputGroupPosition;
}

const mapInputGroupPositionToCSSClass = {
    [EInputGroupPosition.LEFT]: 'cssClass[left]',
    [EInputGroupPosition.INTERMEDIATE]: 'cssClass[intermediate]',
    [EInputGroupPosition.RIGHT]: 'cssClass[right]',
};

/** Компонент ввода суммы. */
export const AmountInput = React.forwardRef<HTMLInputElement, IAmountInputProps>(
    ({className, 'data-test-id': dataTestId, currency, groupPosition, ...rest}, ref) => {
        const {value, disabled, error} = rest;
        const inputRef = useRef<HTMLInputElement | null>(null);

        const renderCurrency = () => (
            <Text
                className="cssClass[currency]"
                data-test-id={dataTestId ? `${dataTestId}__unit` : undefined}
                size={ETextSize.B1}
                type={EFontType.SECONDARY}
                onClick={() => inputRef.current?.focus()}
            >
                {currency}
            </Text>
        );

        /** Функция для хранения ссылки. */
        const setRef = (instance: HTMLInputElement | null) => {
            inputRef.current = instance;
            if (typeof ref === 'function') {
                ref(instance);
            } else if (ref) {
                ref.current = instance;
            }
        };

        return (
            <div
                className={classnames(
                    'cssClass[amountInput]',
                    {
                        'cssClass[filled]': !!value,
                        'cssClass[withCurrency]': !!currency,
                        'cssClass[disabled]': !!disabled,
                        'cssClass[error]': !!error,
                        'cssClass[grouped]': !!groupPosition,
                    },
                    className
                )}
            >
                <AmountBaseInput
                    className={groupPosition && mapInputGroupPositionToCSSClass[groupPosition]}
                    data-test-id={dataTestId && `${dataTestId}__input`}
                    {...rest}
                    ref={setRef}
                />
                {currency && renderCurrency()}
            </div>
        );
    }
);

AmountInput.displayName = 'AmountInput';
