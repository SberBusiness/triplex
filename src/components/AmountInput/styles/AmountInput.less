@import 'src/styles/style';

.amountInput {
    display: flex;

    input {
        text-align: right;
    }

    &.withCurrency input {
        padding-right: @metric-8;
        border-right: none;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    :focus + .currency {
        border-color: var(--triplex-Input-BorderColor_Focus);
    }

    &.filled .currency {
        color: var(--triplex-Input-Color_Default);
    }

    &.disabled .currency {
        cursor: default;
        user-select: none;
        color: var(--triplex-Input-Color_Disabled);
        background-color: var(--triplex-Input-Background_Disabled);
        border-color: var(--triplex-Input-BorderColor_Disabled);
    }

    &.error .currency {
        border-color: var(--triplex-Input-BorderColor_Error);
    }
}

.currency {
    height: @ufs-theme-input-height;
    padding-top: @metric-8 - 1;
    padding-right: @metric-12 - 1;
    padding-bottom: @metric-8 - 1;
    box-sizing: border-box;
    font-size: 14px;
    line-height: normal; // DCBSWT-1866 выравнивание наименования валюты
    white-space: nowrap;
    color: var(--triplex-Input-Placeholder_Color);
    background-color: var(--triplex-Input-Background_Default);
    border: 1px solid var(--triplex-Input-BorderColor_Default);
    border-left-width: 0;
    border-top-right-radius: @ufs-theme-input-border-radius;
    border-bottom-right-radius: @ufs-theme-input-border-radius;
}

.grouped {
    flex: 1;

    &:not(:first-child) {
        margin-left: -1px;
    }

    .left,
    .left + .currency {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    .intermediate,
    .intermediate + .currency {
        border-radius: 0;
    }

    .right,
    .right + .currency {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }

    &.error input,
    &.error input + .currency {
        z-index: 1;
    }

    input:focus,
    input:focus + .currency {
        z-index: 2;
    }
}
