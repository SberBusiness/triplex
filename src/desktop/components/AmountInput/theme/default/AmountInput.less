/* stylelint-disable no-descending-specificity */
@import '../../../../common/theme/default';

.amountInput {
    display: flex;
    align-items: center;

    &.filled .currencyInputRightPart .currencyText {
        color: @color-graphite-01;
    }

    &.error {
        .currencyInputRightPart {
            border-color: @color-ruby-02;
        }
    }

    &.disabled .currencyInputRightPart {
        cursor: default;
        user-select: none;
        background-color: @color-asphalt-05;

        .currencyText {
            color: @color-asphalt-01;
        }
    }

    .currencyInputRightPart {
        display: flex;
        flex-direction: column;
        justify-content: center;
        box-sizing: border-box;
        height: @ufs-theme-input-height;
        border: 1px solid @color-asphalt-03;
        border-radius: 0 @ufs-theme-input-border-radius @ufs-theme-input-border-radius 0;
        border-left: none;
        background-color: @color-snowy;

        .currencyText {
            padding-right: @metric-12;
            padding-left: @metric-4;
            .b2-text-controls-normal(@color-asphalt-01);

            line-height: normal; // DCBSWT-1866 выравнивание наименования валюты
        }
    }

    input {
        padding-right: @metric-4;
        border-right: none;
        border-radius: @ufs-theme-input-border-radius 0 0 @ufs-theme-input-border-radius;
        text-align: right;

        &:focus + .currencyInputRightPart {
            border-color: @color-graphite-01;
        }
    }
}
