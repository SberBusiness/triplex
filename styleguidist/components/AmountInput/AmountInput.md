```jsx
const [value, setValue] = React.useState('');

const maxIntegerDigits = 16;
const fractionDigits = 2;
const currency = 'RUB';

<AmountInput
    value={value}
    maxIntegerDigits={maxIntegerDigits}
    fractionDigits={fractionDigits}
    currency={currency}
    onChange={setValue}
/>
```

### Disabled state

```jsx
const [value, setValue] = React.useState('');

const maxIntegerDigits = 16;
const fractionDigits = 2;
const currency = 'RUB';

<AmountInput
    value={value}
    maxIntegerDigits={maxIntegerDigits}
    fractionDigits={fractionDigits}
    currency={currency}
    onChange={setValue}
    disabled
/>
```

### Error state

```jsx
const [value, setValue] = React.useState('');

const maxIntegerDigits = 16;
const fractionDigits = 2;
const currency = 'RUB';

<AmountInput
    value={value}
    maxIntegerDigits={maxIntegerDigits}
    fractionDigits={fractionDigits}
    currency={currency}
    onChange={setValue}
    error
/>
```
