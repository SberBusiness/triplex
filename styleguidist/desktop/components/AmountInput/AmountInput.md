```jsx
import {ExampleControlPanel} from '../common/ExampleControlPanel/ExampleControlPanel';

const [value, setValue] = React.useState('');
const [disabled, setDisabled] = React.useState(false);
const [error, setError] = React.useState(false);

const maxLength = 19;
const fractionLength = 2;
const currency = 'RUB';

const renderControlPanel = () => (
    <ExampleControlPanel>
        <input
            type="checkbox"
            checked={disabled}
            onChange={(event) => setDisabled(event.target.checked)}
            data-label="Disabled"
        />
        <input
            type="checkbox"
            checked={error}
            onChange={(event) => setError(event.target.checked)}
            data-label="Error"
        />
    </ExampleControlPanel>
);

<>
    {renderControlPanel()}
    <AmountInput
        value={value}
        onChange={setValue}
        disabled={disabled}
        error={error}
        maxLength={maxLength}
        fractionLength={fractionLength}
        currency={currency}
    />
</>
 ```