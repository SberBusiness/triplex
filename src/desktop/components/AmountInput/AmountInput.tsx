import * as React from 'react';
import {IAmountInputProps} from '@sberbusiness/triplex/desktop/components/AmountInput/types';
import {NumberInput} from '@sberbusiness/triplex/desktop/components/NumberInput/NumberInput';

/** Компонент ввода суммы. */
export class AmountInput extends React.PureComponent<IAmountInputProps> {
    public static displayName = 'AmountInput';

    public static defaultProps = {
        fractionLength: 2,
    };

    public render(): JSX.Element {
        const {currency, fractionLength, ...restProps} = this.props;

        return (
            <NumberInput valueFractionLength={fractionLength} placeholderFractionLength={fractionLength} unit={currency} {...restProps} />
        );
    }
}
