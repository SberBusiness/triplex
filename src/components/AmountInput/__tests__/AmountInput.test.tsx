import React from 'react';
import {allure} from '@jest/unit/allure-report';
import renderer from 'react-test-renderer';
import {AmountInput} from '../AmountInput';

describe('AmountInput', () => {
    beforeEach(() => {
        allure.feature('AmountInput');
    });

    it('renders correctly', () => {
        const component = <AmountInput value="1234567890" currency="RUB" onChange={() => void 0} />;
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
