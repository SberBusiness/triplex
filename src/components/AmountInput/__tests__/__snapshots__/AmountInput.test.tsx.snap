// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput renders correctly 1`] = `
<div
  className="cssClass[amountInput] cssClass[filled] cssClass[withCurrency]"
>
  <input
    autoComplete="off"
    className="cssClass[input]"
    data-tx="10.0.0"
    onChange={[Function]}
    onKeyDown={[Function]}
    onSelect={[Function]}
    placeholder="0,00"
    value="1 234 567 890,00"
  />
  <span
    className="cssClass[currency]"
    onClick={[Function]}
  >
    RUB
  </span>
</div>
`;
